### ISSUE TITLE
[out_loki] Records silently dropped

### ISSUE THREAD

**Issue by sean-nixon** [fluent/fluent-bit#4205]
## Bug Report

**Describe the bug**

When using the Loki output, logs seem to be silently dropped with certain configurations. They aren't being delivered to Loki and `proc_records` records for the output does not match the count of records as displayed via the metrics API . Best I cn tell, this seems to be related to using the Loki output together with other outputs. If I configure S3 or Elasticsearch output on it's own or at the same time, all records are successfully sent to S3 / Elasticsearch and `proc_records` for that output match the `records` of the associated input. When using a single Loki output and nothing else, all logs are processed. Adding just the S3 output together with the Loki output results in dropped logs again. The values of `dropped_records`, `errors`, `retries`, `retries_failed`, etc. are all 0. No errors are shown in the logs. I tried enabling trace/debug logging but with my limited understanding of the fluent-bit internals, I could not spot an obvious issue.

**To Reproduce**
- Install td-agent-bit v1.8.8 on Ubuntu 20.04
- Configure systemd input, s3 output, and loki output (Grafana Cloud)

**Expected behavior**
All logs should be processed and sent to both Loki and S3 / Elasticsearch. Calling metrics server at localhost:2020/api/v1/metrics should show the `proc_records` of the S3 output and Loki output matching each other and the value of the input `records`

**Screenshots**
Not a screenshot, but here is what I see in the metrics API.
```
$ curl -s localhost:2020/api/v1/metrics | jq
{
  "input": {
    "system": {
      "records": 200,
      "bytes": 148499
    },
    "storage_backlog.1": {
      "records": 0,
      "bytes": 0
    }
  },
  "filter": {},
  "output": {
    "s3_all": {
      "proc_records": 200,
      "proc_bytes": 148499,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    },
    "loki_system": {
      "proc_records": 114,
      "proc_bytes": 88702,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    }
  }
}
```
**Your Environment**

* Version used: 1.8.8
* Configuration:
```
[SERVICE]
    flush        5
    daemon       Off
    log_level    error
    parsers_file parsers.conf
    plugins_file plugins.conf
    storage.metrics on
    storage.path /var/log/flb-storage
    storage.checksum On
    http_server  On
    http_listen  0.0.0.0
    http_port    2020

[INPUT]
    Name systemd
    Alias system
    tag	 system.*
    DB   /opt/td-agent-bit/systemd.db
    Strip_Underscores On
    Mem_Buf_Limit 79M

[OUTPUT]
    Name  s3
    Alias s3_all
    match *
    bucket <bucket>
    region us-west-2
    s3_key_format /$TAG[0]/stage/$TAG[1]/us-west-2/%Y/%m/%d/${HOSTNAME}_%H%M%S
    canned_acl bucket-owner-full-control
    Retry_Limit False
    storage.total_limit_size 79M

[OUTPUT]
    Name loki
    Alias loki_system
    Match *
    host  <host>
    port  443
    tls On
    http_user <user>
    http_passwd <pass>
    labels source=system
    storage.total_limit_size 79M
```
* Environment name and version (e.g. Kubernetes? What version?): td-agent-bit running on EC2, version 1.8.8
* Server type and version: Grafana Cloud Loki instance
* Operating System and version: Ubuntu 20.04
* Filters and plugins: 
  * Inputs: systemd
  * Outputs: loki, s3
  * Filters: None

**Comment by maciekm** [comment]
I've hit the same issue, missing logs in loki:
```json
  "output": {
    "gelf.0": {
      "proc_records": 358,
      "proc_bytes": 370336,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    },
    "s3.1": {
      "proc_records": 358,
      "proc_bytes": 370336,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    },
    "loki.2": {
      "proc_records": 152,
      "proc_bytes": 160784,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    }
```

When I remove all other outputs I can see more logs in loki but start getting in fluent-bit:
```
entry with timestamp 2021-10-19 22:44:26.292675678 +0000 UTC ignored, reason: 'entry out of order' for stream: {app="app-name", container="container-name", environment="env-name", namespace="namespace-name", node="node-name", pod="pod-name", stream="stderr"},
total ignored: 1 out of 1

[2021/10/19 23:09:09] [engine] caught signal (SIGSEGV)
#0  0x56178abef38d      in  pack_labels() at plugins/out_loki/loki.c:358
#1  0x56178abf16a4      in  loki_compose_payload() at plugins/out_loki/loki.c:1139
#2  0x56178abf17df      in  cb_loki_flush() at plugins/out_loki/loki.c:1177
#3  0x56178ab5027c      in  output_pre_cb_flush() at include/fluent-bit/flb_output.h:514
#4  0x56178b04c646      in  co_init() at lib/monkey/deps/flb_libco/amd64.c:117
#5  0xffffffffffffffff  in  ???() at ???:0
```

* Version used: 1.8.8
* Environment name and version: EKS 1.21
* Filters and plugins:
  * inputs: tail, systemd
  * filters: kubernetes, record_modifier, nest, grep, modify

**Comment by maciekm** [comment]
thanks @sean-nixon - so that leaves just the missing/silently dropped logs issue.

**Comment by maciekm** [comment]
I have tried out `loki` 2.4 image from here: https://github.com/grafana/loki/issues/4647#issuecomment-961457109
The issue persist, loki as a single output works as expected (no out of order issue with loki 2.4):
```json
{
  "input": {
    "tail.0": {
      "records": 2854,
      "bytes": 411657,
      "files_opened": 40,
      "files_closed": 1,
      "files_rotated": 0
    }
  },
  "output": {
    "loki.0": {
      "proc_records": 2854,
      "proc_bytes": 2464483,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    }
  }
}
```
But when together with any other output logs are silently dropped:
```json
{
  "input": {
    "tail.0": {
      "records": 1780,
      "bytes": 520023,
      "files_opened": 32,
      "files_closed": 0,
      "files_rotated": 0
    }
  },
  "output": {
    "s3.0": {
      "proc_records": 1780,
      "proc_bytes": 2278901,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    },
    "loki.1": {
      "proc_records": 1476,
      "proc_bytes": 1900202,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    }
  }
}
```
Another thing is that the output fails completely when other output fails (failing gelf output):
```json
{
  "input": {
    "tail.0": {
      "records": 2198,
      "bytes": 434510,
      "files_opened": 37,
      "files_closed": 0,
      "files_rotated": 0
    }
  },
  "output": {
    "gelf.0": {
      "proc_records": 0,
      "proc_bytes": 0,
      "errors": 0,
      "retries": 301,
      "retries_failed": 290,
      "dropped_records": 2111,
      "retried_records": 2189
    },
    "s3.1": {
      "proc_records": 2189,
      "proc_bytes": 2687894,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    },
    "loki.2": {
      "proc_records": 2,
      "proc_bytes": 1930,
      "errors": 0,
      "retries": 0,
      "retries_failed": 0,
      "dropped_records": 0,
      "retried_records": 0
    }
  }
}
```

**Comment by ldb** [comment]
We are experiencing the same issue. We are currently running two outputs, `kinesis_streams` and `loki`. If the pressure on the kinesis output increases, the loki output starts silently dropping logs.
This is _not_ related to out of order logs, as logs are not rejected by Loki, they are not shipped in the first place.

**Comment by gummeah** [comment]
Would be nice to fix this

**Comment by peter-svensson** [comment]
We're having the exact same problem and would be great with a fix for this.

**Comment by patrick-stephens** [comment]
Possibly resolved by this change which is awaiting 2.0 release: https://github.com/fluent/fluent-bit/pull/6136
